### DISCUSSION TITLE
How do I add the program icon to a ctktoplevel

### DISCUSSION THREAD

**Discussion by zSynctic** [TomSchimansky/CustomTkinter#2085] [Q&A] [answered]
Is there a way I can add the original program icon to the ctktoplevel too?

**Comment by Akascape** [comment]
@zSynctic Still a bug not fixed yet. 

You can try this:
```python
self.after(300, lambda: self.iconbitmap("my_icon.ico"))
```

**Reply by zSynctic** [reply]
mb for not marking as an answer, I was so busy lately, anyways it didn't work for me so can u give me one more example. thanks

**Comment by dipeshSam** [comment]
> Is there a way I can add the original program icon to the ctktoplevel too?

@zSynctic, could you please explain what do you mean by "*original program icon*"? Do you want to set the main window's title bar icon to the toplevel window's title bar icon? Elaborate a bit.

**Reply by zSynctic** [reply]
Yeah that's what I meant, I want to use the original program icon on the ctktoplevel

**Reply by dipeshSam** [reply]
> Yeah that's what I meant, I want to use the original program icon on the ctktoplevel

As I saw, you don't want to do it using `.iconbitmap()` method. You want that toplevel window's icons should automatically be set according to main window. Yes???

If yes, I can put my reply here.

**Reply by zSynctic** [reply]
Yeah

**Comment by dipeshSam** [comment]
> Yeah

@zSynctic, As per the current implementation of the `customtkiner`, it does not provide direct way to set the one icon for all the windows. For a large scale project, if you want to fix one icon in all the windows, you can override the `CTk`, and `CTkToplevel` as per the requirements. The idea is quite simple, we have to call the `iconbitmap()` each time, rather, automate it by saving the icon path and using the saved icon path to the new `toplevel` object.

**Video** demonstration of infinite _same iconed_ windows **providing the icon just one time.**
![auto_ctk_toplevel_icon](https://github.com/TomSchimansky/CustomTkinter/assets/72375763/204c0894-49d9-4583-aec8-f0580e710525)

**Code:**
``` Python
from customtkinter import CTk, CTkToplevel, CTkButton

class App(CTk):
    def __init__(self, fg_color: str | tuple[str, str] | None = None, **kwargs):
        super().__init__(fg_color, **kwargs)

    def iconbitmap(self, bitmap=None, default=None):
        """Modified to save the icon's path for later use."""
        self.icon_path: str = bitmap
        return super().iconbitmap(bitmap, default)
    
class AppToplevel(CTkToplevel):
    def __init__(self, *args, fg_color: str | tuple[str, str] | None = None, **kwargs):
        super().__init__(*args, fg_color=fg_color, **kwargs)

        # Checking whether the master has its own icon. If yes, save the path here also.
        if hasattr(self.master, "icon_path"):
            self.icon_path = getattr(self.master, "icon_path")
            self.after(200, lambda: self.iconbitmap(self.icon_path))
        

if __name__ == "__main__":
    def on_click():
        toplvel1 = AppToplevel(app)
        button = CTkButton(toplvel1, text="Open new toplevel", command=lambda: AppToplevel(toplvel1))
        button.place(relx=0.5, anchor="center", rely=0.5)


    app = App()
    app.geometry("500x350")
    app.iconbitmap("theicon.ico")

    button = CTkButton(app, text="Open new window", command=on_click)
    button.place(relx=0.5, anchor="center", rely=0.5)

    app.mainloop()
```
In the code:

- Basically, `CTk()` is named as `App()`, and `CTkToplevel()` is named as `AppToplevel()`.
- Designed to work in **all the cases** no matter user sets its own icon or not.
- Other functionalities are same producing no errors.
- Memory efficient.


**Hope, it will be helpful for you.**
*Happy customtkinter :)*

**Reply by zSynctic** [reply]
didn't work for me for some reason maybe its because I don't create more than 1 class I only have the app class while the ctktoplevel doesn't

**Reply by dipeshSam** [reply]
> didn't work for me for some reason maybe its because I don't create more than 1 class I only have the app class while the ctktoplevel doesn't

If you provide **your code snippets and their results**, it may help in improving our responds in order to sort out your problem.

Best regards.

**Reply by zSynctic** [reply]
Sorry for the late reply I wasn't around my pc, this is a snippet of my code I don't set the icon in the end I set the main window or level at the start of the class in the __init__ 
```
class App(customtkinter.CTk):
   def __init__(self):
        super().__init__()
        self.p1 = ImageTk.PhotoImage(file=resource("../Assets/icon.ico"))
        self.wm_iconbitmap()
        self.iconphoto(True, self.p1)
   def showsett(self):
   "this is the ctktoplevel this is just a basic example/snippet from my code"
if __name__ == "__main__":
    app = App()
    app.attributes("-topmost", False)
    app.resizable(False, False)
    app.update()
    app.start()
```
Thanks.

**Reply by zSynctic** [reply]
So how can I fix this?

**Comment by Anonymous6598** [comment]
it's set automatically. If you have set icon for CTk(), than CTkTopLevel will inherit the icon too.

**Reply by zSynctic** [reply]
Sadly that's not the case as they sais its a bug ctktoplevel doesn't automatically get a icon unless you do the code snippet they told me about but it isn't working for me idk how to use it I need some help

**Reply by Anonymous6598** [reply]
It might be a bug. Try witting like this:
```python
class Top_window(CTkToplevel):
    def __init__(self: typing.Self, *args: typing.Any, fg_color: str | tuple[str, str] | None = None, **kwargs: typing.Any):
        CTkTopLevel.__init__(self, *args, fg_color=fg_color, **kwargs)

        self.iconbitmap(...) #your icon put here

```

**Reply by zSynctic** [reply]
Can't I just do it using my code layout without using it in a class?

**Reply by Anonymous6598** [reply]
You can

**Reply by Anonymous6598** [reply]
```python
class Top_window(CTkToplevel):
    def __init__(self: typing.Self, *args: typing.Any, fg_color: str | tuple[str, str] | None = None, **kwargs: typing.Any):
        CTkTopLevel.__init__(self, *args, fg_color=fg_color, **kwargs)

        self.after(250, self.iconbitmap(...)) # your icon here

``` 
better version

**Reply by Anonymous6598** [reply]
and also works

**Reply by Anonymous6598** [reply]
@zSynctic, have you tried this code?

**Reply by zSynctic** [reply]
Can u change the code to suit my layout without the class?

**Reply by Anonymous6598** [reply]
Оk, I'll trу.

**Reply by Anonymous6598** [reply]
```Python
from customtkinter import *

class App(CTk):
   def __init__(self, *args, **kwargs):
        super().__init__()
        self.iconbitmap("theicon.ico"))
   
   """
   def showsett(self):
         "this is the ctktoplevel this is just a basic example/snippet from my code"
   """

class AppToplevel(CTkToplevel):
    def __init__(self, *args, fg_color: str | tuple[str, str] | None = None, **kwargs):
        CTkTopLevel.__init__(self, *args, fg_color=fg_color, **kwargs)
        
        self.after(250, lambda: self.iconbitmap("theicon.ico"))

if __name__ == "__main__":
    def on_click():
        toplvel1 = AppToplevel(app)
        button = CTkButton(toplvel1, text="Open new toplevel", command=lambda: AppToplevel(toplvel1))
        button.place(relx=0.5, anchor="center", rely=0.5)


    app = App()
    app.geometry("500x350")

    button = CTkButton(app, text="Open new window", command=on_click)
    button.place(relx=0.5, anchor="center", rely=0.5)

    app.mainloop()
```
I hope now it is understandable.

**Reply by Anonymous6598** [reply]
@zSynctic, is this code now better?

**Reply by zSynctic** [reply]
Thanks, but still no because I would have to change the whole layout of my code, if you have seen my code up earlier the ctktoplevel opens in the function showSett in the class app there isn't 2 classes

**Reply by Anonymous6598** [reply, marked as answer]
```python
from customtkinter import *

class App(CTk):
   def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
       
        self.iconbitmap("theicon.ico")
        self.geometry("500x350")
   
        self.button = CTkButton(self, text="Open new toplevel", command=self.showsett)
        self.button.place(relx=0.5, anchor="center", rely=0.5) 

   def showsett(self):
         self.new = CTkToplevel()
         self.after(250, lambda: self.new.iconbitmap("theicon.ico"))

if __name__ == "__main__":
    app = App()
    app.mainloop()
``` 
now it's perfect

**Reply by zSynctic** [reply]
Thanks man, finally it works
